ethcore/src/private_transactions/encryptor.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Encryption providers.
+
+use std::sync::Arc;
+use std::io::Read;
+use std::iter::repeat;
+use std::time::{Instant, Duration};
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+use parking_lot::Mutex;
+use account_provider::AccountProvider;
+use ethereum_types::{H128, Address};
+use ethjson;
+use ethkey::{Signature, Public};
+use ethcrypto;
+use futures::Future;
+use fetch::{Fetch, Method as FetchMethod, Client as FetchClient};
+use bytes::{Bytes, ToPretty};
+use error::{Error as EthcoreError, PrivateTransactionError};
+
+/// Initialization vector length.
+const INIT_VEC_LEN: usize = 16;
+
+/// Duration of storing retrieved keys (in ms)
+const ENCRYPTION_SESSION_DURATION: u64 = 30 * 1000;
+
+/// Trait for encryption/decryption operations.
+pub trait Encryptor: Send + Sync + 'static {
+	/// Generate unique contract key && encrypt passed data. Encryption can only be performed once.
+	fn encrypt(
+		&self,
+		contract_address: &Address,
+		contract_address_signature: &Signature,
```

Comment:
It seems that `contract_address_signature` could be generated inside these methods (`encrypt` and `decrypt`) since both `AccountProvider` and `requester` are passed here. It will decrease number of arguments && localize all `Address` -> `H256` conversions + private key operations.